core/libp2p/muxer/yamux/yamux.hpp

```diff
@@ -3,234 +3,38 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-#ifndef KAGOME_YAMUX_HPP
-#define KAGOME_YAMUX_HPP
+#ifndef KAGOME_YAMUX_IMPL_HPP
+#define KAGOME_YAMUX_IMPL_HPP
 
-#include <functional>
-#include <map>
-#include <queue>
+#include "libp2p/muxer/connection_muxer.hpp"
 
-#include <boost/asio/streambuf.hpp>
-#include <boost/system/error_code.hpp>
-#include "common/buffer.hpp"
-#include "common/logger.hpp"
 #include "libp2p/muxer/yamux/yamux_config.hpp"
-#include "libp2p/muxer/yamux/yamux_stream_parameters.hpp"
-#include "libp2p/stream/stream.hpp"
-#include "libp2p/transport/connection.hpp"
-#include "libp2p/transport/muxed_connection.hpp"
-
-namespace libp2p::stream {
-  class YamuxStream;
-}
 
 namespace libp2p::muxer {
-  struct YamuxFrame;
-
   /**
-   * Implementation of stream multiplexer - connection, which has only one
-   * physical link to another peer, but many logical streams, for example, for
-   * several applications
-   * Read more: https://github.com/hashicorp/yamux/blob/master/spec.md
+   * Strategy to upgrade connections to Yamuxed ones
    */
-  class Yamux : public transport::MuxedConnection,
-                public std::enable_shared_from_this<Yamux> {
+  class Yamux : public StreamMuxer {
    public:
-    using StreamId = uint32_t;
-    using NewStreamHandler =
-        std::function<void(std::unique_ptr<stream::Stream>)>;
-
-    enum class YamuxErrorStream {
-      NO_SUCH_STREAM = 1,
-      NOT_WRITABLE,
-      NOT_READABLE,
-      YAMUX_IS_CLOSED
-    };
-
     /**
-     * Create a new Yamux instance
-     * @param connection, multiplexed by this instance
-     * @param stream_handler - function, which is going to be called, when a new
-     * stream arrives
-     * @param yamux_config to configure this instance
+     * Create a Yamux strategy
+     * @param config, with which it is to be created
      */
-    Yamux(
-        std::shared_ptr<transport::Connection> connection,
-        NewStreamHandler stream_handler, YamuxConfig yamux_config,
-        kagome::common::Logger logger = kagome::common::createLogger("Yamux"));
-
-    Yamux(const Yamux &other) = delete;
-    Yamux &operator=(const Yamux &other) = delete;
-    Yamux(Yamux &&other) noexcept = delete;
-    Yamux &operator=(Yamux &&other) noexcept = delete;
-
-    ~Yamux() override;
-
-    void start() override;
-
-    void stop() override;
-
-    outcome::result<std::unique_ptr<stream::Stream>> newStream() override;
+    explicit Yamux(YamuxConfig config);
 
-    void close() override;
+    Yamux(const Yamux &other) = default;
```

Comment:
Do we really need all these default items? 
Except of destructor, it is required.

Reply:
Rule of five tells us yes, we need. And I see no reason why they should be deleted, so I made them defaulted